design/endpoint-slice-support.md

```diff
@@ -0,0 +1,101 @@
+# Contour k8s endpoint slice support
+
+Contour currently supports Kubernetes endpoints.
+This document proposes a process to enable support for k8s endpoint slices and, eventually, migrate to exclusively using it.
+
+## Goals
+
+- Contour will support k8s endpoint slices.
+- The endpoint slices will be gated by a feature flag in the Contour config.
+- Kubernetes endpoints will still be the default mode of operation.
+
+## Non-goals
+
+- Removal of k8s endpoints.
+- Support advanced features such as topology aware routing/hints offered by k8s endpoints.
+
+## Background
+
+See [#2696](https://github.com/projectcontour/contour/issues/2696)
+
+Before the introduction of the endpoint slices, the only method for a service to expose the IP addresses and ports of the pads it fronted was through Kubernetes endpoints.
+
+However, Kubernetes endpoints have a few challenges with scaling, such as
+- transmission of the entire endpoint object even if one endpoint changes, leading to network contention and heavy recalculation on the consumers of this data.
+- As of k8s v1.22, a hard limit of 1000 addresses per endpoint object.
+- No topology aware information for more intelligent routing.
+
+To solve the above problems, endpoint slices was introduced in 1.16 and, as of k8s v1.21 in [stable state](https://kubernetes.io/docs/concepts/services-networking/endpoint-slices/).
+
+
+## High-Level Design
+
+A feature flag will gate the usage of endpoint slices, and the implementation will use the same pattern as used in the `endpointtranslator.go` with a few differences listed in the detailed design.
+
+## Detailed Design
```

Comment:
Should/will any of the [EndpointSlice conditions](https://kubernetes.io/docs/concepts/services-networking/endpoint-slices/#conditions) be taken into account?

Reply:
I've chosen to stick with just the `ready` condition to keep things simpler for initial implementation and have feature parity with endpoints. Added a section regarding that in https://github.com/projectcontour/contour/pull/5703/commits/63fac731b8b745fbe893ca69e3739cd3b21641f5